kmod/src/data.c

```diff
@@ -1565,48 +1563,95 @@ int scoutfs_data_fiemap(struct inode *inode, struct fiemap_extent_info *fieinfo,
 	iblock = start >> SCOUTFS_BLOCK_SM_SHIFT;
 	last = (start + len - 1) >> SCOUTFS_BLOCK_SM_SHIFT;
 
+	args.ino = ino;
+	args.inode = inode;
+
+	/* outer loop */
 	while (iblock <= last) {
-		ret = scoutfs_ext_next(sb, &data_ext_ops, &args,
-				       iblock, 1, &ext);
-		if (ret < 0) {
-			if (ret == -ENOENT)
-				ret = 0;
-			last_flags = FIEMAP_EXTENT_LAST;
+		/* lock */
+		inode_lock(inode);
+		down_read(&si->extent_sem);
+
+		ret = scoutfs_lock_inode(sb, SCOUTFS_LOCK_READ, 0, inode, &lock);
+		if (ret) {
+			lock = NULL;
 			break;
 		}
 
-		trace_scoutfs_data_fiemap_extent(sb, ino, &ext);
+		args.lock = lock;
 
-		if (ext.start > last) {
-			/* not setting _LAST, it's for end of file */
-			ret = 0;
-			break;
+		/* collect entries */
+		info = page_address(page);
+		while (entries < (PAGE_SIZE / sizeof(struct fiemap_extent)) - 1) {
+			ret = scoutfs_ext_next(sb, &data_ext_ops, &args,
+					       iblock, 1, &ext);
+			if (ret < 0) {
+				if (ret == -ENOENT)
+					ret = 0;
+				last_flags = FIEMAP_EXTENT_LAST;
+				break;
+			}
+
+			trace_scoutfs_data_fiemap_extent(sb, ino, &ext);
+
+			if (ext.start > last) {
+				/* not setting _LAST, it's for end of file */
+				ret = 0;
+				break;
+			}
+
+			if (scoutfs_ext_can_merge(&cur, &ext)) {
+				/* merged extents could be greater than input len */
+				cur.len += ext.len;
+			} else {
+				/* fill it */
+				info->fe_logical = cur.start;
```

Comment:
The idea behind using the kernel fiemap struct in the page was to be able to pass them directly to the kernel's fill extent.  If you're bouncing them back into the scoutfs_extent before calling fill then the page storage might as well just use the scoutfs extents.  (The dissonance that triggers this realization is that it looks weird to store the block granular start in the byte granular fe_logical.)

Reply:
I'll stick it to scoutfs_extent for now, since it's the smaller of the options, resulting in less lock/unlock cycles.